tests/foreman/virtwho/test_api_virtwho.py

```diff
@@ -67,6 +71,26 @@ def _make_virtwho_configure(self):
             args[u'hypervisor_password'] = self.hypervisor_password
         return args
 
+    def _try_to_get_guest_bonus(self, hypervisor_name, sku):
+        vdc_id = None
+        subscriptions = entities.Subscription().search(
+            query={'search': '{0}'.format(sku)})
+        for item in subscriptions:
+            item = item.read_json()
+            if hypervisor_name.lower() in item['hypervisor']['name']:
+                vdc_id = item['id']
+                break
+        return vdc_id
+
+    def _get_guest_bonus(self, hypervisor_name, sku):
+        vdc_id = self._try_to_get_guest_bonus(hypervisor_name, sku)
+        if vdc_id is None:
+            sleep(5)
```

Comment:
In general, explicit sleeps like that are not recommended. What if `vdc_id` could be assigned after 1 second? You just wait for 4 seconds doing nothing. What if `vdc_id` is still not available after 5 seconds, but could be available after 6 seconds? Your test failed, while it probably should not.

You should use the loop instead. Check if value is available, and if not, wait for a second or two. Then check again. If value is available, return it. At the same time, track how much time passed since you started. If you tried enough times, throw exception.

One example of such loop is in https://github.com/SatelliteQE/robottelo/blob/master/robottelo/api/utils.py#L677 .
You can also use [`wait_for`](https://github.com/RedHatQE/wait_for), which we already pull in in Robottelo.

Reply:
Yes, thanks

Reply:
And I like the `wait_for` function, it can also help me resolve the issue regarding getting host failed, thanks again